### ISSUE TITLE
[Milestone 4] 事件驱动决策步与 DQN Credit Assignment 实验

### ISSUE THREAD

**Issue by Firestl** [Firestl/gold-miner-sim#7]
## 目标

Milestone 3 在随机地图上完成了泛化评估，但 200k DQN 得到退化策略：回收结束后倾向立即再次 FIRE，100 个固定 evaluation seeds 上 mean=472，低于 random baseline mean=637。

本阶段只修改 **RL 决策时序**：

- `SWINGING` 时 Agent 才做决策；
- `WAIT` 仍推进固定 10 个 physics ticks；
- `FIRE` 后自动推进完整出钩/回收过程，直到重新进入 `SWINGING` 或 episode 结束；
- 自动推进期间累计 reward，并把结果直接返回给触发该次 FIRE 的 RL transition。

保持随机地图、26 维 observation、reward、action space、physics 和 DQN 配置不变。

目标是验证：减少无效 action step、缩短 FIRE 到 reward 的 credit-assignment 路径后，DQN 是否更容易学到基于 observation 的出钩时机。

---

## 1. 不修改底层 `GoldMinerEnv`

本 Issue 不修改以下契约：

- `GoldMinerEnv` 的 `fixed/random` map mode；
- random spawn points；
- action space：`Discrete(2)`；
- `WAIT=0` / `FIRE=1`；
- observation：26 维 `float32`；
- reward：`new_score - old_score`；
- episode duration：60 s；
- physics：60 Hz；
- Hook 状态机；
- collision；
- object value / radius / retract speed；
- Pygame renderer。

不要通过修改 `env.py` 的 reward、物理或 observation 来改善训练结果。

---

## 2. 保留现有 `DecisionIntervalWrapper`

现有：

```text
DecisionIntervalWrapper
```

继续保留，现有测试继续通过，不删除、不改变语义。

它是 Milestone 2/3 的 baseline 决策模型：每个 RL action 固定推进 10 个 physics ticks。

本 Issue 新增一个 wrapper，不覆盖旧实现。

---

## 3. 新增 `SwingDecisionWrapper`

修改：

```text
src/gold_miner_sim/wrappers.py
```

新增：

```python
class SwingDecisionWrapper(gymnasium.Wrapper):
    ...
```

该 wrapper 是 Gold Miner 专用 wrapper，不需要抽象成通用 event-driven framework。

### 核心不变量

除 episode 已结束外，每次 `SwingDecisionWrapper.step(action)` 返回时，底层 Hook 必须处于：

```text
SWINGING
```

因此 Agent 不会在以下状态做决策：

```text
EXTENDING
RETRACT_EMPTY
RETRACT_LOADED
```

---

## 4. `WAIT` 的精确语义

Agent 在 `SWINGING` 状态选择：

```text
WAIT
```

wrapper 执行最多 10 个底层 physics ticks：

```text
WAIT
WAIT
WAIT
WAIT
WAIT
WAIT
WAIT
WAIT
WAIT
WAIT
```

要求：

1. 每个底层 tick 调用一次 `env.step(WAIT)`；
2. 累计全部 reward；
3. 任意 tick 出现 `terminated=True` 或 `truncated=True` 时立即停止；
4. 正常情况下执行满 10 ticks 后返回；
5. 返回最后一个底层 tick 的 observation / info / flags；
6. 返回累计 reward。

这部分行为应与现有 `DecisionIntervalWrapper.step(WAIT)` 等价。

固定：

```text
WAIT_INTERVAL = 10 physics ticks
```

不要把该值做成 CLI、配置文件或超参数搜索项。

---

## 5. `FIRE` 的精确语义

Agent 在 `SWINGING` 状态选择：

```text
FIRE
```

wrapper 按以下顺序执行：

### 5.1 第一个 tick

执行：

```text
env.step(FIRE)
```

正常情况下 Hook 从：

```text
SWINGING -> EXTENDING
```

累计该 tick reward。

如果该 tick 已使 episode `terminated/truncated`，立即返回。

### 5.2 自动推进

如果 episode 未结束，之后只执行：

```text
env.step(WAIT)
```

持续推进底层 physics，经过：

```text
EXTENDING
    -> RETRACT_EMPTY
    -> SWINGING
```

或：

```text
EXTENDING
    -> RETRACT_LOADED
    -> SWINGING
```

每个底层 tick：

- 累计 reward；
- 检查 `terminated/truncated`；
- 检查 Hook 是否重新进入 `SWINGING`。

满足以下任一条件立即结束本次 wrapper step：

1. Hook 重新进入 `SWINGING`；
2. `terminated=True`；
3. `truncated=True`。

不得在回到 `SWINGING` 后额外执行一个 WAIT tick。

### 5.3 Reward

本次 wrapper step 返回从 FIRE tick 到重新进入 SWINGING/episode end 之间的 **全部 reward 总和**。

例如成功抓回 DIAMOND：

```text
Agent action: FIRE
    ↓
EXTENDING
    ↓
hit diamond
    ↓
RETRACT_LOADED
    ↓
回收到顶部
    ↓
+500
    ↓
SWINGING
```

该次 `SwingDecisionWrapper.step(FIRE)` 应直接返回：

```text
reward = 500
```

空钩完整往返应返回：

```text
reward = 0
```

如果 episode 在回收完成前超时，则未完成物体仍按底层环境规则不计分，wrapper 返回实际累计 reward。

---

## 6. 不改变 reward 定义

这里不是 reward shaping。

底层环境仍然只在物体真正回收到顶部时产生分数 reward：

```text
reward = new_score - old_score
```

wrapper 只是把多个 physics tick 的 reward 累计到一个 RL transition。

不要新增：

- FIRE reward；
- hit reward；
- WAIT penalty；
- empty-shot penalty；
- angle reward；
- target proximity reward。

---

## 7. Variable-duration transition 是本阶段预期行为

`WAIT` 通常消耗：

```text
10 ticks = 1/6 s
```

`FIRE` 会消耗完整一次出钩和回收过程，因此不同 FIRE transition 的 physics duration 可能不同。

这是有意设计。

本阶段不要实现：

- duration-aware discount；
- SMDP 专用算法；
- 根据实际 tick 数修改 `gamma`；
- 自定义 replay buffer。

DQN 的 `gamma` 继续使用 SB3 默认值。

本 Milestone 只验证“折叠非决策状态”这一变量。

---

## 8. Wrapper 状态要求

`SwingDecisionWrapper` 只应在合法 decision point 接受动作。

正常使用流程：

```text
reset()
  -> SWINGING

step(WAIT)
  -> SWINGING or episode end

step(FIRE)
  -> SWINGING or episode end
```

不要允许调用者通过该 wrapper 获得 `EXTENDING/RETRACT_*` 状态下的决策机会。

实现保持简单，不需要增加 `DecisionPoint` enum、状态机框架或 callback。

---

## 9. 训练切换到 `SwingDecisionWrapper`

修改：

```text
scripts/train_dqn.py
```

训练链改为：

```text
GoldMinerEnv(render_mode=None, map_mode="random")
    -> SwingDecisionWrapper
    -> Monitor
    -> DQN
```

不要再在本阶段训练脚本中使用 `DecisionIntervalWrapper`。

### DQN 配置保持 Milestone 3 不变

| 参数 | 值 |
|---|---|
| algorithm | `DQN` |
| policy | `MlpPolicy` |
| total_timesteps | `200_000` |
| seed | `0` |
| buffer_size | `100_000` |
| learning_starts | `1_000` |
| exploration_fraction | `0.3` |
| 其他参数 | SB3 默认 |

不要调：

- learning rate；
- gamma；
- exploration 参数；
- network architecture；
- batch size；
- target update；
- train frequency。

默认输出改为：

```text
models/dqn_gold_miner_swing.zip
runs/dqn_swing/
```

`--seed` 是整个训练实验 seed，SB3 会同时用于模型/探索等随机性以及环境 RNG 的可复现初始化；不要在训练 loop 中手工反复 `reset(seed=...)`。

---

## 10. Random baseline 使用同一 wrapper

修改：

```text
scripts/random_baseline.py
```

环境链改为：

```text
GoldMinerEnv(render_mode=None, map_mode="random")
    -> SwingDecisionWrapper
```

其他评估规则保持 Milestone 3 不变：

```text
episodes = 100
map seeds = 1000..1099
```

每局：

```text
env.reset(seed=map_seed)
```

Random policy 每个 **真实 decision point** 随机选择：

```text
WAIT / FIRE
```

Action-space RNG 整次 baseline 只 seed 一次，不要每局重置成相同 action sequence。

输出保持：

```text
Episodes
Seed range
Mean score
Std score
Min score
Max score
```

Milestone 3 的 random baseline 637 不能直接作为本阶段 baseline；wrapper 改变后必须重新测量 random baseline。

---

## 11. DQN evaluation 使用同一 wrapper

修改：

```text
scripts/eval_dqn.py
```

环境链改为：

```text
GoldMinerEnv(map_mode="random")
    -> SwingDecisionWrapper
```

Headless 默认仍评估：

```text
100 episodes
seeds 1000..1099
```

使用：

```python
model.predict(obs, deterministic=True)
```

输出保持：

```text
Episodes
Seed range
Mean score
Std score
Min score
Max score
```

默认模型：

```text
models/dqn_gold_miner_swing.zip
```

### Pygame replay

继续支持：

```text
--render --seed <map_seed>
```

render 模式默认 1 episode。

人工观察至少 3 个不同 seed，例如：

```text
1000
1007
1042
```

重点观察：

- Agent 是否会在 SWINGING 中等待；
- FIRE angle 是否随地图变化；
- 是否仍然每次回收结束立即 FIRE；
- 是否会针对不同布局产生明显不同的 action timing。

---

## 12. 最低测试

新增 `SwingDecisionWrapper` 测试。

不要在 pytest 中运行 DQN training。

至少覆盖：

### 12.1 WAIT 等价性

一次：

```text
SwingDecisionWrapper.step(WAIT)
```

应与裸环境手工执行 10 次：

```text
WAIT x 10
```

结果一致。

比较至少：

- observation；
- reward；
- hook state；
- angle；
- rope length；
- remaining time。

### 12.2 FIRE 空钩

构造一个确定会空钩的发射场景。

调用一次：

```text
wrapper.step(FIRE)
```

必须：

- 内部完成 EXTENDING -> RETRACT_EMPTY -> SWINGING；
- 返回时 hook state 为 SWINGING；
- reward=0；
- rope length 回到 minimum；
- remaining time 按实际底层 ticks 正确减少。

### 12.3 FIRE 命中并回收

构造一个确定会命中物体的场景。

调用一次：

```text
wrapper.step(FIRE)
```

必须：

- 内部经过 RETRACT_LOADED；
- 返回时已重新进入 SWINGING；
- 返回 reward 等于该物体 value；
- score 已增加；
- object 已 inactive；
- attached object 已清空。

### 12.4 Episode timeout during auto-advance

构造 episode 剩余时间不足以完成当前 FIRE 的情况。

调用一次 FIRE 后：

- wrapper 在 `truncated=True` 的底层 tick 立即返回；
- 不继续推进；
- remaining time=0；
- 未完整回收到顶部的物体不加分；
- 不要求返回状态为 SWINGING，因为 episode 已结束。

### 12.5 整局 reward 一致

通过 `SwingDecisionWrapper` 跑完整 episode：

```text
sum(wrapper rewards) == final env score
```

### 12.6 旧 wrapper 回归

现有 `DecisionIntervalWrapper` 全部测试继续通过。

---

### A. Tests

```bash
uv run pytest -q
```

必须全部通过。

### B. 新 random baseline

```bash
uv run --group train python scripts/random_baseline.py --episodes 100 --seed 1000
```

记录 mean / std / min / max。

### C. Train

```bash
uv run --group train python scripts/train_dqn.py --timesteps 200000 --seed 0
```

必须生成：

```text
models/dqn_gold_miner_swing.zip
runs/dqn_swing/
```

### D. Headless evaluation

```bash
uv run --group train python scripts/eval_dqn.py --episodes 100 --seed 1000
```

记录 mean / std / min / max。

### E. Visual evaluation

至少执行：

```bash
uv run --group train python scripts/eval_dqn.py --seed 1000 --render
uv run --group train python scripts/eval_dqn.py --seed 1007 --render
uv run --group train python scripts/eval_dqn.py --seed 1042 --render
```

---

## 14. PR 必须记录的实验结果

PR 描述必须包含：

### Random baseline

```text
Episodes
Seed range
Mean
Std
Min
Max
```

### DQN training

```text
Timesteps
Training seed
Episodes（Monitor 可得时记录）
模型路径
```

### DQN evaluation

```text
Episodes
Seed range
Mean
Std
Min
Max
```

### Comparison

至少记录：

```text
mean_random_swing
mean_dqn_swing
delta = mean_dqn_swing - mean_random_swing
```

并与 Milestone 3 现象做一句定性对比：

```text
旧策略是否“回收后立即 FIRE”
新策略是否出现可观察的 WAIT / map-dependent FIRE timing
```

不要为了让 delta 为正而修改本 Issue 明确固定的训练参数。

---

## 15. 学习效果判断

本阶段希望观察到：

```text
mean_dqn_swing > mean_random_swing
```

以及不同 map seed 下 FIRE timing 有明显变化。

这两个是实验目标，不是 pytest 条件，也不是代码合并硬门槛。

如果 DQN 仍然低于 random baseline：

- 不延长训练；
- 不调超参数；
- 不 reward shaping；
- 不切 PPO；
- 不修改随机地图；
- 在 PR 中记录结果与行为现象，下一 Milestone 再分析。

---

## 16. 明确不做

本 Issue 不实现：

- PPO / A2C；
- DQN 超参数调优；
- hyperparameter sweep；
- reward shaping；
- action masking 库；
- 修改 observation；
- 修改 random spawn points；
- 修改 decision interval=10 的 WAIT cadence；
- duration-aware/SMDP discount；
- prioritized replay；
- n-step return；
- custom DQN；
- vectorized env；
- TensorBoard/W&B；
- 真实游戏 calibration；
- RGB/YOLO；
- 新物体/炸药/商店。

---

## Definition of Done

- [ ] 新增 `SwingDecisionWrapper`；
- [ ] 保留 `DecisionIntervalWrapper` 原行为；
- [ ] Agent 只在 SWINGING 状态获得决策机会；
- [ ] WAIT 固定推进最多 10 physics ticks；
- [ ] FIRE 自动推进直到重新 SWINGING 或 episode end；
- [ ] wrapper 正确累计自动推进期间 reward；
- [ ] 空钩 / 命中回收 / timeout 行为有测试；
- [ ] wrapper episode reward 总和等于 final score；
- [ ] 现有全部环境和旧 wrapper 测试继续通过；
- [ ] training 改用 random map + `SwingDecisionWrapper`；
- [ ] baseline/eval 使用相同 wrapper 与 seeds 1000..1099；
- [ ] 完成一次默认 200k DQN 训练；
- [ ] 完成 100-map random baseline 与 DQN evaluation；
- [ ] 可用 Pygame 对指定 seed 回放模型；
- [ ] PR 记录 baseline、DQN、delta 和行为观察；
- [ ] 未通过调参/reward shaping/PPO 等方式补救实验结果。
